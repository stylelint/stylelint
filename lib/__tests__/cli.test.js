/* eslint no-console: off */

'use strict';

const path = require('path');

const cli = require('../cli');
const pkg = require('../../package.json');
const replaceBackslashes = require('../testUtils/replaceBackslashes');

const fixturesPath = replaceBackslashes(path.join(__dirname, 'fixtures'));
const { buildCLI } = cli;

jest.mock('get-stdin');

describe('buildCLI', () => {
	it('flags - default', () => {
		expect(buildCLI([]).flags).toEqual({
			allowEmptyInput: false,
			cache: false,
			color: false,
			disableDefaultIgnores: false,
			fix: false,
			formatter: 'string',
			help: false,
			ignoreDisables: false,
			printConfig: false,
			quiet: false,
			reportInvalidScopeDisables: false,
			reportNeedlessDisables: false,
			stdin: false,
			version: false,
		});
	});

	it('flags.allowEmptyInput', () => {
		expect(buildCLI(['--allow-empty-input']).flags.allowEmptyInput).toBe(true);
		expect(buildCLI(['--aei']).flags.allowEmptyInput).toBe(true);
	});

	it('flags.cache', () => {
		expect(buildCLI(['--cache']).flags.cache).toBe(true);
	});

	it('flags.cacheLocation', () => {
		expect(buildCLI(['--cache-location=foo']).flags.cacheLocation).toBe('foo');
	});

	it('flags.color', () => {
		expect(buildCLI(['--color']).flags.color).toBe(true);
		expect(buildCLI(['--no-color']).flags.color).toBe(false);
	});

	it('flags.config', () => {
		expect(buildCLI(['--config=/path/to/file']).flags.config).toBe('/path/to/file');
	});

	it('flags.configBasedir', () => {
		expect(buildCLI(['--config-basedir=/path/to/dir']).flags.configBasedir).toBe('/path/to/dir');
	});

	it('flags.customFormatter', () => {
		expect(buildCLI(['--custom-formatter=foo']).flags.customFormatter).toBe('foo');
	});

	it('flags.customSyntax', () => {
		expect(buildCLI(['--custom-syntax=foo']).flags.customSyntax).toBe('foo');
	});

	it('flags.disableDefaultIgnores', () => {
		expect(buildCLI(['--disable-default-ignores']).flags.disableDefaultIgnores).toBe(true);
		expect(buildCLI(['--di']).flags.disableDefaultIgnores).toBe(true);
	});

	it('flags.fix', () => {
		expect(buildCLI(['--fix']).flags.fix).toBe(true);
	});

	it('flags.formatter', () => {
		expect(buildCLI(['--formatter=json']).flags.formatter).toBe('json');
		expect(buildCLI(['-f', 'json']).flags.formatter).toBe('json');
	});

	it('flags.help', () => {
		expect(buildCLI(['--help']).flags.help).toBe(true);
		expect(buildCLI(['-h']).flags.help).toBe(true);
	});

	it('flags.ignoreDisables', () => {
		expect(buildCLI(['--ignore-disables']).flags.ignoreDisables).toBe(true);
		expect(buildCLI(['--id']).flags.ignoreDisables).toBe(true);
	});

	it('flags.ignorePath', () => {
		expect(buildCLI(['--ignore-path=/path/to/file']).flags.ignorePath).toBe('/path/to/file');
		expect(buildCLI(['-i', '/path/to/file']).flags.ignorePath).toBe('/path/to/file');
	});

	it('flags.ignorePattern', () => {
		expect(buildCLI(['--ignore-pattern=vendor/**']).flags.ignorePattern).toEqual(['vendor/**']);
		expect(buildCLI(['--ip', 'vendor/**']).flags.ignorePattern).toEqual(['vendor/**']);
		expect(buildCLI(['--ip', 'vendor/**', '--ip', 'test/*.css']).flags.ignorePattern).toEqual([
			'vendor/**',
			'test/*.css',
		]);
	});

	it('flags.maxWarnings', () => {
		expect(buildCLI(['--max-warnings=7']).flags.maxWarnings).toBe(7);
		expect(buildCLI(['--mw', '-1']).flags.maxWarnings).toBe(-1);
	});

	it('flags.outputFile', () => {
		expect(buildCLI(['--output-file=/path/to/file']).flags.outputFile).toBe('/path/to/file');
		expect(buildCLI(['-o', '/path/to/file']).flags.outputFile).toBe('/path/to/file');
	});

	it('flags.printConfig', () => {
		expect(buildCLI(['--print-config']).flags.printConfig).toBe(true);
	});

	it('flags.quiet', () => {
		expect(buildCLI(['--quiet']).flags.quiet).toBe(true);
		expect(buildCLI(['-q']).flags.quiet).toBe(true);
	});

	it('flags.reportInvalidScopeDisables', () => {
		expect(buildCLI(['--report-invalid-scope-disables']).flags.reportInvalidScopeDisables).toBe(
			true,
		);
		expect(buildCLI(['--risd']).flags.reportInvalidScopeDisables).toBe(true);
	});

	it('flags.reportNeedlessDisables', () => {
		expect(buildCLI(['--report-needless-disables']).flags.reportNeedlessDisables).toBe(true);
		expect(buildCLI(['--rd']).flags.reportNeedlessDisables).toBe(true);
	});

	it('flags.stdin', () => {
		expect(buildCLI(['--stdin']).flags.stdin).toBe(true);
	});

	it('flags.stdinFilename', () => {
		expect(buildCLI(['--stdin-filename=foo.css']).flags.stdinFilename).toBe('foo.css');
	});

	it('flags.syntax', () => {
		expect(buildCLI(['--syntax=less']).flags.syntax).toBe('less');
		expect(buildCLI(['-s', 'less']).flags.syntax).toBe('less');
	});

	it('flags.version', () => {
		expect(buildCLI(['--version']).flags.version).toBe(true);
		expect(buildCLI(['-v']).flags.version).toBe(true);
	});
});

describe('CLI', () => {
	beforeAll(() => {
		jest.spyOn(process, 'exit').mockImplementation(() => {});
		jest.spyOn(process.stdout, 'write').mockImplementation(() => {});
		jest.spyOn(process.stderr, 'write').mockImplementation(() => {});
		jest.spyOn(console, 'log').mockImplementation(() => {});
		jest.spyOn(console, 'error').mockImplementation(() => {});
	});

	afterAll(() => {
		jest.restoreAllMocks();
	});

	it('basic', async () => {
		await cli([]);

		expect(process.exit).toHaveBeenCalledWith(2);

		expect(console.log).toHaveBeenCalledTimes(1);
		expect(console.log).toHaveBeenCalledWith(
			expect.stringContaining('Usage: stylelint [input] [options]'),
		);
	});

	it('--help', async () => {
		await cli(['--help']);

		expect(process.exit).toHaveBeenCalledWith(0);

		expect(console.log).toHaveBeenCalledTimes(1);
		expect(console.log.mock.calls[0][0]).toMatchSnapshot();
	});

	it('--version', async () => {
		await cli(['--version']);

		expect(process.exitCode).toBeUndefined();

		expect(console.log).toHaveBeenCalledTimes(1);
		expect(console.log).toHaveBeenCalledWith(expect.stringContaining(pkg.version));
	});

	it('--print-config', async () => {
		await cli([
			'--print-config',
			'--config',
			`${fixturesPath}/config-color-no-invalid-hex.json`,
			`${fixturesPath}/invalid-hex.css`,
		]);

		expect(process.exitCode).toBeUndefined();

		expect(process.stdout.write).toHaveBeenCalledTimes(1);
		expect(process.stdout.write).toHaveBeenLastCalledWith(
			JSON.stringify(
				{
					rules: {
						'color-no-invalid-hex': [true],
					},
				},
				null,
				'  ',
			),
		);
	});

	it('--report-needless-disables', async () => {
		await cli([
			'--report-needless-disables',
			'--config',
			replaceBackslashes(path.join(fixturesPath, 'config-block-no-empty.json')),
			replaceBackslashes(path.join(fixturesPath, 'empty-block-with-disables.css')),
		]);

		expect(process.exitCode).toBe(2);

		expect(process.stdout.write).toHaveBeenCalledTimes(2);
		expect(process.stdout.write).toHaveBeenNthCalledWith(
			1,
			expect.stringContaining('unused rule: color-named'),
		);
		expect(process.stdout.write).toHaveBeenNthCalledWith(
			2,
			expect.stringContaining('Unexpected empty block'),
		);
	});

	it('--stdin', async () => {
		await cli(['--stdin', '--config', `${fixturesPath}/config-no-empty-source.json`]);

		expect(process.exitCode).toBe(2);

		expect(process.stdout.write).toHaveBeenCalledTimes(1);
		expect(process.stdout.write).toHaveBeenNthCalledWith(
			1,
			expect.stringContaining('Unexpected empty source'),
		);
	});

	it('exits with non zero on unfound module in config', async () => {
		await cli([
			'--config',
			replaceBackslashes(path.join(fixturesPath, 'config-require-unknown.js')),
			replaceBackslashes(path.join(fixturesPath, 'empty-block-with-disables.css')),
		]);

		expect(process.exitCode).toEqual(1);

		expect(process.stderr.write).toHaveBeenCalledTimes(1);
		expect(process.stderr.write).toHaveBeenCalledWith(
			expect.stringContaining('Cannot find module'),
		);
	});
});
