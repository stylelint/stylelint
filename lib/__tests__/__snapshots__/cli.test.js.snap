// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CLI --help 1`] = `
"
  A mighty, modern CSS linter.

  Usage: stylelint [input] [options]

  Input: Files(s), glob(s), or nothing to use stdin.

    If an input argument is wrapped in quotation marks, it will be passed to
    globby for cross-platform glob support. node_modules are always ignored.
    You can also pass no input and use stdin, instead.

  Options:

    --config

      Path to a specific configuration file (JSON, YAML, or CommonJS), or the
      name of a module in node_modules that points to one. If no --config
      argument is provided, stylelint will search for configuration files in
      the following places, in this order:
        - a stylelint property in package.json
        - a .stylelintrc file (with or without filename extension:
          .json, .yaml, .yml, and .js are available)
        - a stylelint.config.js file exporting a JS object
      The search will begin in the working directory and move up the directory
      tree until a configuration file is found.

    --config-basedir

      An absolute path to the directory that relative paths defining \\"extends\\"
      and \\"plugins\\" are *relative to*. Only necessary if these values are
      relative paths.

    --print-config

      Print the configuration for the given path.

    --ignore-path, -i

      Path to a file containing patterns that describe files to ignore. The
      path can be absolute or relative to process.cwd(). By default, stylelint
      looks for .stylelintignore in process.cwd().

    --ignore-pattern, --ip

      Pattern of files to ignore (in addition to those in .stylelintignore)

    --syntax, -s

      Specify a syntax. Options: \\"css\\", \\"css-in-js\\", \\"html\\", \\"less\\",
      \\"markdown\\", \\"sass\\", \\"scss\\", \\"sugarss\\". If you do not specify a syntax,
      syntaxes will be automatically inferred by the file extensions
      and file content.

    --fix

      Automatically fix violations of certain rules.

    --custom-syntax

      Module name or path to a JS file exporting a PostCSS-compatible syntax.

    --stdin

      Accept stdin input even if it is empty.

    --stdin-filename

      A filename to assign stdin input.

    --ignore-disables, --id

      Ignore styleline-disable comments.

    --disable-default-ignores, --di

      Allow linting of node_modules.

    --cache                       [default: false]

      Store the info about processed files in order to only operate on the
      changed ones the next time you run stylelint. By default, the cache
      is stored in \\"./.stylelintcache\\". To adjust this, use --cache-location.

    --cache-location              [default: '.stylelintcache']

      Path to a file or directory to be used for the cache location.
      Default is \\"./.stylelintcache\\". If a directory is specified, a cache
      file will be created inside the specified folder, with a name derived
      from a hash of the current working directory.

      If the directory for the cache does not exist, make sure you add a trailing \\"/\\"
      on *nix systems or \\"\\\\\\" on Windows. Otherwise the path will be assumed to be a file.

    --formatter, -f               [default: \\"string\\"]

      The output formatter: \\"compact\\", \\"json\\", \\"string\\", \\"unix\\" or \\"verbose\\".

    --custom-formatter

      Path to a JS file exporting a custom formatting function.

    --quiet, -q

      Only register violations for rules with an \\"error\\"-level severity (ignore
      \\"warning\\"-level).

    --color
    --no-color

      Force enabling/disabling of color.

    --report-needless-disables, --rd

      Also report errors for stylelint-disable comments that are not blocking a lint warning.
      The process will exit with code 2 if needless disables are found.

    --report-invalid-scope-disables, --risd

      Report stylelint-disable comments that used for rules that don't exist within the configuration object.
      The process will exit with code 2 if invalid scope disables are found.

    --max-warnings, --mw

      Number of warnings above which the process will exit with code 2.
      Useful when setting \\"defaultSeverity\\" to \\"warning\\" and expecting the
      process to fail on warnings (e.g. CI build).

    --output-file, -o

      Path of file to write report.

    --version, -v

      Show the currently installed version of stylelint.

    --allow-empty-input, --aei

      When glob pattern matches no files, the process will exit without throwing an error.
"
`;
